packages/esm-amd-loader/src/esm-amd-loader.ts

```diff
@@ -18,42 +18,349 @@ interface Window {
+    args,
+    deps: depModules,
+    moduleBody,
+  };
+  return stateTransition(module, stateData);
+}
+
+function loadDeps(
+    module: Module, depSpecifiers: string[]): [Array<{}>, Module[]] {
+  const args: Array<{}> = [];
+  const depModules: Module[] = [];
+  for (const depSpec of depSpecifiers) {
+    if (depSpec === 'exports') {
+      args.push(module.exports);
+      continue;
+    }
+    if (depSpec === 'require') {
+      args.push(function(
+          deps: string[],
+          onResolve?: ResolveCallback,
+          onError?: ErrorCallback) {
+        const [args, depModules] = loadDeps(module, deps);
+
+        waitOnDeps(depModules, () => {
+          if (onResolve) {
+            onResolve.apply(null, args);
+          }
+        }, onError);
+      });
+      continue;
+    }
+    if (depSpec === 'meta') {
+      args.push({
+        // We append "#<script index>" to top-level scripts so that they have
+        // unique keys in the registry. We don't want to see that here.
+        url: (module.isTopLevel === true) ? baseUrl : module.url
+      });
+      continue;
+    }
+
+    // We have a dependency on a real module.
+    const dependency = getModule(resolveUrl(module.urlBase, depSpec));
+    args.push(dependency.exports);
+    depModules.push(dependency);
+
+    if (dependency.stateData.state === StateEnum.Initialized) {
+      load(dependency as Module<Initialized>);
+    }
+  }
+  return [args, depModules];
+}
+
+/**
+ * Start executing our dependencies, in order, as they become available.
+ * Once they're all executed, execute our own module body, if any.
+ */
+function beginWaitingOnDeps(module: Module<WaitingOnEarlierScripts>) {
+  const stateData: WaitingOnDeps = {
+    state: StateEnum.WaitingOnDeps,
+    args: module.stateData.args,
+    deps: module.stateData.deps,
+    moduleBody: module.stateData.moduleBody
+  };
+  const mutatedModule = stateTransition(module, stateData);
+  waitOnDeps(
+      module.stateData.deps,
+      () => execute(mutatedModule),
+      (e) => fail(mutatedModule, e));
+  return mutatedModule;
+}
+
+/** Runs the given module body. */
+function execute(module: Module<WaitingOnDeps>): Module<Executed|Failed> {
+  const stateData = module.stateData;
+  if (stateData.moduleBody != null) {
+    try {
+      stateData.moduleBody.apply(null, stateData.args);
+    } catch (e) {
+      return fail(module, e);
+    }
+  }
+  return stateTransition(module, {state: StateEnum.Executed});
+}
+
+/**
+ * Called when a module has failed to load, either becuase its script errored,
+ * or because one of its transitive dependencies errored.
+ */
+function fail(mod: Module, error: Error) {
+  return stateTransition(mod, {state: StateEnum.Failed, error});
+}
+
+/**
+ * Transition the given module to the given state.
+ *
+ * Does not do any checking that the transition is legal.
+ * Calls onNextStateChange callbacks.
+ */
+function stateTransition<SD extends ModuleStateData>(
```

Comment:
Move this up higher? Feels like this should be close to the state enum since it's so general to the state machine.

Reply:
+1, done